### PR TITLE
[PoS] Stake Modifier V2

### PR DESCRIPTION
**Pull request by random-zebra**
This pull request defines a new 256-bits modifier for the proof of stake protocol, `CBlockIndex::nStakeModifierV2`. 
It is computed at every block, by taking the hash of the modifier of previous block along with the coinstake input.
To meet the protocol, the PoS kernel must comprise the modifier of the previous block.

Enforcement height is set at 1214000 for testnet and 1967000 for mainnet.

This also includes upgrades to `PROTOCOL_VERSION` (to 70917) and to block version (to version 6) and disables 0-fee transactions.

### CODE REVIEW

**Comment by CaveSpectre11** [comment]
@random-zebra evaluate this to see if you think it would make sense to cherry pick into this?

https://github.com/CaveSpectre11/PIVX/commit/3fcb1b367665698c1b81cc9c35d90319d31a968e

**Comment by CaveSpectre11** [comment]
> @random-zebra evaluate the below commit, and see if would be appropriate to cherry pick in?
> 
> [CaveSpectre11@3fcb1b3](https://github.com/CaveSpectre11/PIVX/commit/3fcb1b367665698c1b81cc9c35d90319d31a968e)

Nevermind.   routine will be dormant now to just chain validation

File: src/main.cpp [hunk 1]
```diff
@@ -5051,30 +5041,6 @@ void CBlockIndex::BuildSkip()
...
-            }
-        }
-    }
-    if (nMints || nSpends)
-        LogPrintf("%s : block contains %d zPIV mints and %d zPIV spends\n", __func__, nMints, nSpends);
-
-    if (!CheckBlockSignature(*pblock))
-        return error("ProcessNewBlock() : bad proof-of-stake block signature");
-
     if (pblock->GetHash() != Params().HashGenesisBlock() && pfrom != NULL) {
```
Comment (CaveSpectre11) [inline src/main.cpp]: another nullptr change.

File: src/main.cpp [hunk 2]
```diff
@@ -5084,17 +5050,26 @@ bool ProcessNewBlock(CValidationState& state, CNode* pfrom, CBlock* pblock, CDis
...
     }
 
+    // Preliminary checks
+    int64_t nStartTime = GetTimeMillis();
+
+    // check block
+    bool checked = CheckBlock(*pblock, state);
+
+    if (!CheckBlockSignature(*pblock))
+        return error("ProcessNewBlock() : bad proof-of-stake block signature");
```
Comment (CaveSpectre11) [inline src/main.cpp]: __func__

Comment (furszy) [reply src/main.cpp]: Hey, based on the importance of this PR, let's try to focus the review toward the real purpose and not clean main.cpp big scary file.

The file should be cleaned and divided in several files in several PRs, similar organization to what bitcoin has now (would be good to have an open issue for this work).

Comment (CaveSpectre11) [reply src/main.cpp]: fair enough; just catching things in code that was touched given similar things were changed in this PR as well.  (And it was initial feedback given time constraints before I dug in deeper).  But of course it's one of the reasons I make them as comments only.  

However... msg received :)

**Comment by Mrs-X** [comment]
On a first view looks good, even when 
```
    // switch with old modifier on upgrade block
    if (pindexPrev->nStakeModifierV2 == uint256() && pindexPrev->nStakeModifier)
```

(in my opinion) is unnecessary hard to verify/understand (you have to lookup those assumptions in chain.h) than simply using the block height of the upgrade block.

File: src/kernel.cpp [hunk 3]
```diff
@@ -255,90 +250,134 @@ bool GetKernelStakeModifier(uint256 hashBlockFrom, uint64_t& nStakeModifier, int
         nStakeModifier = pindexFrom->nStakeModifier;
         return true;
     }
-    int64_t nStakeModifierSelectionInterval = GetStakeModifierSelectionInterval();
     const CBlockIndex* pindex = pindexFrom;
-    CBlockIndex* pindexNext = chainActive[pindexFrom->nHeight + 1];
+    CBlockIndex* pindexNext = chainActive[pindex->nHeight + 1];;
```
Comment (JSKitty) [inline src/kernel.cpp]: Double semicolon.

**Pull request merged**